File: net/JetBrains.Serialization/src/Dataholders/GeneralNameInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using JetBrains.Serialization;
using Org.BouncyCastle.Asn1.X509;

[JsonObject(MemberSerialization.OptIn)]
public class GeneralNameInfo : IEncodableInfo
{
  [JsonProperty("Name")] private X509NameInfo _name;

  [JsonProperty("Tag")] private int _tag;

  [JsonConstructor]
  public GeneralNameInfo(X509NameInfo name, int tag)
  {
    _name = name;
    _tag = tag;
  }

  public GeneralNameInfo(GeneralName generalName)
  {
    _name = new X509NameInfo((X509Name)generalName.Name);
    _tag = generalName.TagNo;
  }

  public Asn1Encodable ToPrimitive()
  {
    return TaggedObjectInfo.GetTaggedObject(_tag == 4, _tag, _name.ToPrimitive());
  }
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/GeneralNameInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.x500.X500Name
import org.bouncycastle.asn1.x509.GeneralName

@Serializable
data class GeneralNameInfo(
  val name: X500NameInfo,
  val tag: Int
) : EncodableInfo {

  constructor(generalName: GeneralName) : this(
    X500NameInfo(generalName.name as X500Name),
    generalName.tagNo
  )

  override fun toPrimitive(): ASN1Primitive =
    TaggedObjectInfo.getTaggedObject(
      tag == 4,
      tag,
      name.toPrimitive()
    )
}